### ISSUE TITLE
Question: how can the empty type {} be detected?

### ISSUE THREAD

**Issue by aleksey-bykov** [Microsoft/TypeScript#3620]
I wonder what the sufficient set of checks is that detects the empty type as a result of failed type inference. I have an expression and a type checker that gives me its type and symbol. My problem is that I get false positives on module references and type parameters.

**Comment by RyanCavanaugh** [comment]
It would be useful to post at least one code example of a thing you'd like to detect

**Comment by aleksey-bykov** [comment]
Like here https://github.com/Microsoft/TypeScript/issues/3423

**Comment by aleksey-bykov** [comment]
```typescript
function f<a>(): a { return undefined; }
var value = f(); // <-- value is {}
```

**Comment by aleksey-bykov** [comment]
my clueless attempt to code it

```typescript

function isExpressionOfEmptyType(expression: ts.Expression, checker: ts.TypeChecker) : boolean {
    var typeOpt = checker.getTypeAtLocation(expression);
    return typeOpt != null && isEmptyType(typeOpt);
}
function isEmptyType(type: ts.Type) : boolean {
    return isObjectLike(type)
        && !hasMembers(type)
        && !isTypeParameter(type);
}
var objectLikeTypeFlags = ts.TypeFlags.ContainsObjectLiteral & ts.TypeFlags.ObjectLiteral & ts.TypeFlags.Anonymous;
function isObjectLike(type: ts.Type) : boolean {
    return (type.flags & objectLikeTypeFlags) === objectLikeTypeFlags;
}
function hasMembers(type: ts.Type) : boolean {
    return type.getCallSignatures().length > 0 || type.getProperties().length > 0 || type.getConstructSignatures().length > 0;
}
function isTypeParameter(type: ts.Type) : boolean {
    return (type.flags & ts.TypeFlags.TypeParameter) === ts.TypeFlags.TypeParameter;
}

```

**Comment by DanielRosenwasser** [comment]
This was how I did it: https://github.com/Microsoft/TypeScript/commit/abd44b1d7ec4d3d546a4ce7a2b47729892c7014a

**Comment by aleksey-bykov** [comment]
It helped thanks
